File: CS50/credit/credit.c
#include <cs50.h>
#include <stdio.h>

long get_cardnumber(void);
void check_card(long n);
bool luhn_algorithm(long n);

//get card number from user and check if it is amex, mastercard, visa or invalid card
int main(void)
{
    long cardnumber = get_cardnumber();
    check_card(cardnumber);
}

//prompt user for cardnumber
long get_cardnumber(void)
{
    long l = get_long("Number: ");
    return l;
}

//check if the card is amex
bool is_amex(long n)
{
    int i = n / 10000000000000;
    if (i == 34 || i == 37)
    {
        if (luhn_algorithm(n))
        {
            return true;
        }
    }
    else
    {
        return false;
    }
    return false;
}

//check if the card is mastercard
bool is_mastercard(long n)
{
    int i = n / 100000000000000;
    if (i >= 51 && i <= 55 && luhn_algorithm(n))
    {
        if (luhn_algorithm(n))
        {
            return true;
        }
    }
    else
    {
        return false;
    }
    return false;
}

//check if the card is visa
bool is_visa(long n)
{
    int i = n / 1000000000000;
    int j = n / 1000000000000000;

    if (i == 4 || j == 4)
    {
        if (luhn_algorithm(n))
        {
            return true;
        }
    }
    else
    {
        return false;
    }
    return false;
}

//check if the card is valid using luhn's algorithm
bool luhn_algorithm(long n)
{
    int sum = 0;
    int i = 0;
    while (n > 0)
    {
        if (i % 2 == 0)
        {
            sum = sum + n % 10;
        }
        else
        {
            if ((n % 10) < 5)
            {
                sum = sum + (n % 10) * 2;
            }
            else
            {
                sum = sum + n % 10 * 2 / 10 + n % 10 * 2 % 10;
            }
        }
        n = n / 10;
        i++;
    }
    if (sum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

//call the methods to check if the card is amex, mastercard, visa or invalid
void check_card(long n)
{
    if (is_amex(n))
    {
        printf("AMEX\n");
    }
    else if (is_mastercard(n))
    {
        printf("MASTERCARD\n");
    }
    else if (is_visa(n))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: CS50/sentimental-credit/credit.py
# TODO
from cs50 import get_string
import re


def is_valid(number):
    l = len(number)
    total = 0
    times2 = False
    for i in range(l):
        n = int(number[l - 1 - i])
        if times2:
            if n * 2 < 10:
                total += 2 * n
            else:
                total += ((2 * n) % 10 + int((2 * n) / 10))
        else:
            total += n
        times2 = not times2
    if total % 10 == 0:
        return True
    return False


number = get_string("Number: ")

if len(re.findall('\d', number)) == len(number):

    if len(number) == 15 and (int(number[0:2]) == 34 or int(number[0:2]) == 37) and is_valid(number):
        print('AMEX')

    if len(number) == 16 and int(number[0:2]) in range(51, 56) and is_valid(number):
        print('MASTERCARD')

    if (len(number) == 13 or len(number) == 16) and int(number[0]) == 4 and is_valid(number):
        print('VISA')

    print('INVALID')

else:
    print('INVALID')
